test/ICSharpCode.SharpZipLib.Tests/Zip/ZipEncryptionHandling.cs

```diff
@@ -108,6 +120,101 @@ public void ZipFileAesDecryption()
 			}
 		}
 
+		/// <summary>
+		/// Tests for reading encrypted entries using ZipInputStream.
+		/// </summary>
+		[Test]
+		[Category("Encryption")]
+		[Category("Zip")]
+		public void ZipInputStreamDecryption(
+			[Values(0, 128, 256)] int aesKeySize,
+			[Values(CompressionMethod.Stored, CompressionMethod.Deflated)]
+			CompressionMethod compressionMethod,
+			[Values] bool forceDataDescriptor)
+		{
+			ZipInputStreamDecryption(aesKeySize, compressionMethod, forceDataDescriptor, false, false);
+		}
+		
+		/// <summary>
+		/// Tests for reading encrypted entries using ZipInputStream.
+		/// Verify that it is possible to skip reading of entries.
+		/// </summary>
+		[Test]
+		[Category("Encryption")]
+		[Category("Zip")]
+		public void ZipInputStreamDecryptionSupportsSkippingEntries(
+			[Values(0, 128, 256)] int aesKeySize,
+			[Values(CompressionMethod.Stored, CompressionMethod.Deflated)]
+			CompressionMethod compressionMethod,
+			[Values] bool forceDataDescriptor)
+		{
+			ZipInputStreamDecryption(aesKeySize, compressionMethod, forceDataDescriptor, true, false);
+		}		
+		
+		
+		/// <summary>
+		/// Tests for reading encrypted entries using ZipInputStream.
+		/// Verify that it is possible to read entries only partially.
+		/// </summary>
+		[Test]
+		[Category("Encryption")]
+		[Category("Zip")]
+		public void ZipInputStreamDecryptionSupportsPartiallyReadingOfEntries(
+			[Values(0, 128, 256)] int aesKeySize,
+			[Values(CompressionMethod.Stored, CompressionMethod.Deflated)]
+			CompressionMethod compressionMethod,
+			[Values] bool forceDataDescriptor)
+		{
+			ZipInputStreamDecryption(aesKeySize, compressionMethod, forceDataDescriptor, false, true);
+		}			
+
+		private void ZipInputStreamDecryption(
+			[Values(0, 128, 256)]int aesKeySize, 
+			[Values(CompressionMethod.Stored, CompressionMethod.Deflated)]CompressionMethod compressionMethod, 
+			[Values]bool forceDataDescriptor, 
+			[Values]bool skipEntries, 
+			[Values]bool partiallyReadEntries)
+		{
+			var password = "password";
+
+			using (var ms = forceDataDescriptor ? new MemoryStreamWithoutSeek() : new MemoryStream())
+			{
+				WriteEncryptedZipToStream(ms, 3, 3, password, aesKeySize, compressionMethod);
+				ms.Seek(0, SeekOrigin.Begin);
+
+				using (var zis = new ZipInputStream(ms))
+				{
+					zis.IsStreamOwner = false;
+					zis.Password = password;
+
+					for (int i = 0; i < 6; i++)
+					{
+						var entry = zis.GetNextEntry();
+						int fileNumber = int.Parse(entry.Name[5].ToString());
+
+						if (skipEntries && fileNumber % 2 == 1)
+						{
+							continue;
+						}
+						
+						if (partiallyReadEntries && fileNumber % 2 == 1)
+						{
+							zis.ReadByte();
+							continue;
+						}
```

Comment:
I think you are missing the point a little. This part is not good for a test. You have several tests in the same body and and switching them out.
And I still have no idea why you are checking "short" entries for some tests and not for others.
If some tests have a common setup/teardown, that part *can* be moved to a helper function, but it's also fine to just have several copies of the same code.
The test should clearly test for a single thing, not require you to resolve the values of bools to check. This means that the reviewers and subsequent developers can look at the test and directly see what the scope and purpose of the test is.

Also, as a general rule, don't have bools at all in helpers, and if you do, always call them with the name (like the StreamReader just below). That way you don't have to consult the docs or look up the function to understand what the purpose is.

Reply:
Split into several tests.

I check here for different sizes of entries as I figured out that the behavior internally is different in case of AES depending if the entry fits into the buffer of the InflaterInputStream or not. If I have it correctly in mind then it is about the calculation of the AES Auth code.